text/0000-ptr-to-field.md

```diff
@@ -0,0 +1,210 @@
+- Feature Name: ptr-to-field
+- Start Date: 2019-06-05
+- RFC PR: [rust-lang/rfcs#0000](https://github.com/rust-lang/rfcs/pull/0000)
+- Rust Issue: [rust-lang/rust#0000](https://github.com/rust-lang/rust/issues/0000)
+
+# Summary
+[summary]: #summary
+
+This feature could serve as the backbone for some pointer to field syntax, and even if no syntax is made, this feature serves as a safe generic way to talk about types and their fields.
+
+# Motivation
+[motivation]: #motivation
+
+The motivation for this feature is to allow safe projection through smart pointers, for example `Pin<&mut T>` to `Pin<&mut Field>`. This is a much needed feature to make `Pin<P>` more usable in safe-contexts, without the need to use unsafe to map to a field. This also can allow projection through other smart pointers like `Rc<T>`, `Arc<T>`. This feature cannot be implemented as a library effectively because it depends on the layouts of types, so it requires integration with the Rust compiler until Rust gets a stable layout (which may never happen).
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+First the core traits, type, and functions that need to be added
+
+```rust
+/// Contains metadata about how to get to the field from the parent using raw pointers
+/// This is an opaque type that never needs to be stablized, and it only an implementation detail
+struct FieldDescriptor {
+    ...
+}
+
+/// The compiler should prevent user implementations for `Field`,
+/// i.e. only the compiler is allowed to implement `Field`
+/// This is to prevent people from creating fake "fields"
+unsafe trait Field {
+    /// The type that the field is a part of
+    type Parent;
+    /// The type of the field
+    type Type;
+
+    /// The metadata required to get to the field using raw pointers
+    const FIELD_DESCRIPTOR: FieldDescriptor;
+}
+
+trait Project<F: Field> {
+    /// The projected version of Self
+    type Projection;
+
+    fn project(self, field: F) -> Self::Projection;
+}
+
+impl<T: ?Sized> *const T {
+    unsafe fn project_unchecked<F: Field<Parent = T>>(self, field: F) -> *const F::Type {
+        // make the field pointer, this code is allowed to assume that
+        // self points to a valid instance of T
+        ... 
+    }
+}
+
+impl<T: ?Sized> *mut T {
+    unsafe fn project_unchecked<F: Field<Parent = T>>(self, field: F) -> *mut F::Type {
+        // make the field pointer, this code is allowed to assume that
+        // self points to a valid instance of T
+        ...
+    }
+}
+```
+
+Now we need some syntax to refer to the fields types. Some ideas for the syntax are
+
+* `Parent.field`  // my favorite, as it seems most natural
```

Comment:
Does this work for tuple structs?

```rust
struct S(u8, u16);
let field = S.0; // yes?
```

Reply:
Yes, this should work with tuple structs, and tuples as well.

`<(u8, u16)>.1 == u16`

Reply:
Why not for `union`? Unlike `enum` there are no representational issues with the discriminant. Every union variant already has a concrete type behaviour is defined only when manually checking the initialized portions anyways so I don't see why it should not be possible. Note that this is vitally important for `MaybeUninit`.

Reply:
Because unions would allow safely going form `&union` to `&field` using `Project`, but this is unsafe. I don't want to have to make a zoo of traits to turn on or off `Project` so `union`s can be punted to another RFC.

Reply:
You're right, too bad. I would have expected it to be possible with the same `unsafe` operations as raw pointers at least. This could however work if the projection of `&_` is not guaranteed available for all fields, e.g. requires an additional `unsafe` trait.

Reply:
I added `union` support by making it unsafe to construct the field types, but this is tenuous (and may be removed before it is accepted). No other type requires unsafe to construct, they may use privacy to force you to use unsafe functions to construct them. But the actual construction is always considered safe.

`enums` will remain unsupported.